src/rgw/rgw_ioc_dispatch.cc

```diff
@@ -0,0 +1,270 @@
+    auto ec = boost::system::error_code{-ret, boost::system::system_category()};
+    ceph::async::post(std::move(p), ec, bufferlist{});
+  } else {
+    p.release();
+  }
+  return init.result.get();
+}
+
+namespace {
+  struct Handler {
+  rgw::Aio* throttle = nullptr;
+  rgw::AioResult& r;
+  // read callback
+  void operator()(boost::system::error_code ec, bufferlist bl) const {
+    r.result = -ec.value();
+    r.data = std::move(bl);
+    throttle->put(r);
+  }
+};
+
+void file_aio_read_abstract(boost::asio::io_context& context, spawn::yield_context yield,
+                            std::string& file_path, off_t read_ofs, off_t read_len,
+                            rgw::Aio* aio, rgw::AioResult& r) {
+  using namespace boost::asio;
+  async_completion<spawn::yield_context, void()> init(yield);
+  auto ex = get_associated_executor(init.completion_handler);
+  async_read(context, file_path, read_ofs, read_len, bind_executor(ex, Handler{aio,r}));
+}
+} //anonymous namespace
+
+void file_aio_read_abstract(optional_yield y, std::string& file_path,
+                           off_t read_ofs, off_t read_len,
+                           rgw::Aio* aio, rgw::AioResult& r) {
+  file_aio_read_abstract(y.get_io_context(), y.get_yield_context(),
+                         file_path, read_ofs, read_len, aio, r);
+}
+
+}//namespace ioc
+
+
+
+void IOChook::init(CephContext* _cct, Context* on_finish) {
+  m_cct = _cct;
+  ceph_assert(m_cct != nullptr);
+  std::unique_lock locker{m_lock};
+  if(m_cache_client == nullptr) {
+    auto controller_path = m_cct->_conf.template get_val<std::string>(
+      "immutable_object_cache_sock");
+    m_cache_client = new CacheClient(controller_path.c_str(), m_cct);
+  }
+  create_cache_session(on_finish, false);
+}
+
+void IOChook::create_cache_session(Context *on_finish, bool is_reconnect) {
+	ceph_assert(ceph_mutex_is_locked_by_me(m_lock));
+  if(m_connecting) {
+    return;
+  }
+
+  m_connecting = true;
+
+  Context* register_ctx = new LambdaContext([this, on_finish](int ret) {
+    if(ret < 0) {
+      ldout(m_cct, 1) << "rgw_hook: IOC hook fail to register client." << dendl;
+    }
+    handle_register_client(ret < 0 ? false : true);
+
+    ceph_assert(m_connecting);
+    m_connecting = false;
+
+    if(on_finish != nullptr) {
+      on_finish->complete(0);
+    }
+  });
+
+  Context* connect_ctx = new LambdaContext([this, register_ctx](int ret) {
+    if(ret < 0) {
+      ldout(m_cct, 1)  << "rgw_hook: IOC hook fail to connect to IOC daemon" << dendl;
+      register_ctx->complete(0);
+      return;
+    }
+
+    ldout(m_cct, 20) << "rgw_hook: IOC hook connected to IOC daemon." << dendl;
+
+    m_cache_client->register_client(register_ctx);
+  });
+
+
+	if(m_cache_client != nullptr && is_reconnect) {
+    delete m_cache_client;
+    auto controller_path = m_cct->_conf.template get_val<std::string>(
+            "immutable_object_cache_sock");
+    m_cache_client = new CacheClient(controller_path.c_str(), m_cct);
```

Comment:
indentation?

Reply:
Does it means this?
```c++
    delete m_cache_client;
    auto controller_path = m_cct->_conf.template get_val<std::string>(
      "immutable_object_cache_sock");
    m_cache_client = new CacheClient(controller_path.c_str(), m_cct);
```
if it is OK, resolved.

Reply:
Oh, Sorry, I just found out that the indentation you mentioned is caused by tab。